File: flutter_ui/lib/models/job.dart
class Job {
  final String id;
  final String status;
  final String prompt;
  final List<String> skillIds;
  final List<String> skillTags;
  final String workingDir;
  final String? model;
  final double? maxBudgetUsd;
  final List<String>? allowedTools;
  final Map<String, dynamic>? outputDest;
  final String? source;
  final int priority;
  final List<String> tags;
  final String createdAt;
  final String? startedAt;
  final String? completedAt;
  final String? workerId;
  final String? error;
  final double? costUsd;
  final int? durationMs;
  final int retryCount;
  final int? timeoutSecs;
  final String? cronId;
  final dynamic skillSnapshot;
  final String? assembledPrompt;

  Job({
    required this.id,
    required this.status,
    required this.prompt,
    this.skillIds = const [],
    this.skillTags = const [],
    this.workingDir = '.',
    this.model,
    this.maxBudgetUsd,
    this.allowedTools,
    this.outputDest,
    this.source,
    this.priority = 5,
    this.tags = const [],
    required this.createdAt,
    this.startedAt,
    this.completedAt,
    this.workerId,
    this.error,
    this.costUsd,
    this.durationMs,
    this.retryCount = 0,
    this.timeoutSecs,
    this.cronId,
    this.skillSnapshot,
    this.assembledPrompt,
  });

  factory Job.fromJson(Map<String, dynamic> json) => Job(
        id: json['id'] ?? '',
        status: json['status'] ?? 'unknown',
        prompt: json['prompt'] ?? '',
        skillIds: List<String>.from(json['skill_ids'] ?? []),
        skillTags: List<String>.from(json['skill_tags'] ?? []),
        workingDir: json['working_dir'] ?? '.',
        model: json['model'],
        maxBudgetUsd: (json['max_budget_usd'] as num?)?.toDouble(),
        allowedTools: json['allowed_tools'] != null
            ? List<String>.from(json['allowed_tools'])
            : null,
        outputDest: json['output_dest'] is Map<String, dynamic>
            ? json['output_dest']
            : null,
        source: json['source'],
        priority: json['priority'] ?? 5,
        tags: List<String>.from(json['tags'] ?? []),
        createdAt: json['created_at'] ?? '',
        startedAt: json['started_at'],
        completedAt: json['completed_at'],
        workerId: json['worker_id'],
        error: json['error'],
        costUsd: (json['cost_usd'] as num?)?.toDouble(),
        durationMs: json['duration_ms'] as int?,
        retryCount: json['retry_count'] ?? 0,
        timeoutSecs: json['timeout_secs'] as int?,
        cronId: json['cron_id'],
        skillSnapshot: json['skill_snapshot'],
        assembledPrompt: json['assembled_prompt'],
      );

  String get shortId => id.length >= 8 ? id.substring(0, 8) : id;

  String get promptPreview =>
      prompt.length > 80 ? '${prompt.substring(0, 80)}...' : prompt;
}

class JobResult {
  final String jobId;
  final String result;
  final double costUsd;
  final int durationMs;

  JobResult({
    required this.jobId,
    required this.result,
    required this.costUsd,
    required this.durationMs,
  });

  factory JobResult.fromJson(Map<String, dynamic> json) => JobResult(
        jobId: json['job_id'] ?? '',
        result: json['result'] ?? '',
        costUsd: (json['cost_usd'] as num?)?.toDouble() ?? 0,
        durationMs: json['duration_ms'] as int? ?? 0,
      );
}

class QueueStatus {
  final int pending;
  final int running;
  final int completed;
  final int failed;

  QueueStatus({
    this.pending = 0,
    this.running = 0,
    this.completed = 0,
    this.failed = 0,
  });

  factory QueueStatus.fromJson(Map<String, dynamic> json) => QueueStatus(
        pending: json['pending'] ?? 0,
        running: json['running'] ?? 0,
        completed: json['completed'] ?? 0,
        failed: json['failed'] ?? 0,
      );
}

File: crates/claw-models/src/job.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Job {
    pub id: Uuid,
    pub status: JobStatus,
    pub prompt: String,
    #[serde(default)]
    pub skill_ids: Vec<String>,
    #[serde(default)]
    pub skill_tags: Vec<String>,
    #[serde(default)]
    pub tool_ids: Vec<String>,
    #[serde(default = "default_working_dir")]
    pub working_dir: PathBuf,
    pub model: Option<String>,
    pub max_budget_usd: Option<f64>,
    pub allowed_tools: Option<Vec<String>>,
    #[serde(default)]
    pub output_dest: OutputDest,
    #[serde(default)]
    pub source: JobSource,
    #[serde(default = "default_priority")]
    pub priority: u8,
    #[serde(default)]
    pub tags: Vec<String>,
    pub created_at: DateTime<Utc>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub worker_id: Option<String>,
    pub error: Option<String>,
    pub cost_usd: Option<f64>,
    pub duration_ms: Option<u64>,
    #[serde(default)]
    pub retry_count: u32,
    pub timeout_secs: Option<u64>,
    pub workspace_id: Option<Uuid>,
    pub cron_id: Option<Uuid>,
    pub template_id: Option<Uuid>,
    pub pipeline_run_id: Option<Uuid>,
    pub pipeline_step: Option<usize>,
    pub skill_snapshot: Option<serde_json::Value>,
    pub assembled_prompt: Option<String>,
}

fn default_priority() -> u8 {
    5
}

fn default_working_dir() -> PathBuf {
    PathBuf::from(".")
}

/// Request body for creating a new job via the API.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateJobRequest {
    pub prompt: String,
    #[serde(default)]
    pub skill_ids: Vec<String>,
    #[serde(default)]
    pub skill_tags: Vec<String>,
    #[serde(default)]
    pub tool_ids: Vec<String>,
    pub working_dir: Option<PathBuf>,
    pub model: Option<String>,
    pub max_budget_usd: Option<f64>,
    pub allowed_tools: Option<Vec<String>>,
    #[serde(default)]
    pub output_dest: OutputDest,
    #[serde(default)]
    pub tags: Vec<String>,
    pub priority: Option<u8>,
    pub timeout_secs: Option<u64>,
    pub workspace_id: Option<Uuid>,
    pub template_id: Option<Uuid>,
}

/// Summary returned after submitting a job.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateJobResponse {
    pub id: Uuid,
    pub status: JobStatus,
    pub created_at: DateTime<Utc>,
}

/// Response for job result queries.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobResultResponse {
    pub job_id: Uuid,
    pub result: String,
    pub cost_usd: f64,
    pub duration_ms: u64,
}

/// Queue status overview.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueueStatus {
    pub pending: u64,
    pub running: u64,
    pub completed: u64,
    pub failed: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, strum::Display, strum::EnumString)]
#[strum(serialize_all = "lowercase")]
#[serde(rename_all = "lowercase")]
pub enum JobStatus {
    Pending,
    Running,
    Completed,
    Failed,
    Cancelled,
}

impl Default for JobStatus {
    fn default() -> Self {
        Self::Pending
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum OutputDest {
    Redis,
    File { path: PathBuf },
    Webhook { url: String },
}

impl Default for OutputDest {
    fn default() -> Self {
        Self::Redis
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, strum::Display, strum::EnumString)]
#[strum(serialize_all = "lowercase")]
#[serde(rename_all = "lowercase")]
pub enum JobSource {
    Cli,
    Api,
    Cron,
    #[strum(serialize = "filewatcher")]
    FileWatcher,
}

impl Default for JobSource {
    fn default() -> Self {
        Self::Api
    }
}
